// @threshold-network/solidity-contracts/export/deploy/07_deploy_token_staking.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var hardhat_1 = require("hardhat");
var func = function (hre) {
    return __awaiter(this, void 0, void 0, function () {
        var getNamedAccounts, deployments, execute, log, deployer, T, KeepTokenStaking, NuCypherStakingEscrow, VendingMachineKeep, VendingMachineNuCypher, KeepStake, tokenStakingConstructorArgs, tokenStakingInitializerArgs, tokenStakingAddress, TokenStaking, tokenStaking, implementationInterface, jsonAbi, tokenStakingDeployment, fs, TokenStaking;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    getNamedAccounts = hre.getNamedAccounts, deployments = hre.deployments;
                    execute = deployments.execute, log = deployments.log;
                    return [4 /*yield*/, getNamedAccounts()];
                case 1:
                    deployer = (_a.sent()).deployer;
                    return [4 /*yield*/, deployments.get("T")];
                case 2:
                    T = _a.sent();
                    return [4 /*yield*/, deployments.get("KeepTokenStaking")];
                case 3:
                    KeepTokenStaking = _a.sent();
                    return [4 /*yield*/, deployments.get("NuCypherStakingEscrow")];
                case 4:
                    NuCypherStakingEscrow = _a.sent();
                    return [4 /*yield*/, deployments.get("VendingMachineKeep")];
                case 5:
                    VendingMachineKeep = _a.sent();
                    return [4 /*yield*/, deployments.get("VendingMachineNuCypher")];
                case 6:
                    VendingMachineNuCypher = _a.sent();
                    return [4 /*yield*/, deployments.get("KeepStake")];
                case 7:
                    KeepStake = _a.sent();
                    tokenStakingConstructorArgs = [
                        T.address,
                        KeepTokenStaking.address,
                        NuCypherStakingEscrow.address,
                        VendingMachineKeep.address,
                        VendingMachineNuCypher.address,
                        KeepStake.address,
                    ];
                    tokenStakingInitializerArgs = [];
                    if (!(hre.network.name == "mainnet")) return [3 /*break*/, 10];
                    return [4 /*yield*/, hardhat_1.ethers.getContractFactory("TokenStaking")];
                case 8:
                    TokenStaking = _a.sent();
                    return [4 /*yield*/, hardhat_1.upgrades.deployProxy(TokenStaking, tokenStakingInitializerArgs, {
                            constructorArgs: tokenStakingConstructorArgs,
                        })];
                case 9:
                    tokenStaking = _a.sent();
                    tokenStakingAddress = tokenStaking.address;
                    log("Deployed TokenStaking with TransparentProxy at ".concat(tokenStakingAddress));
                    implementationInterface = tokenStaking.interface;
                    jsonAbi = implementationInterface.format(hardhat_1.ethers.utils.FormatTypes.json);
                    tokenStakingDeployment = {
                        address: tokenStakingAddress,
                        abi: JSON.parse(jsonAbi),
                    };
                    fs = require("fs");
                    fs.writeFileSync("TokenStaking.json", JSON.stringify(tokenStakingDeployment, null, 2), "utf8", function (err) {
                        if (err) {
                            console.log(err);
                        }
                    });
                    log("Saved TokenStaking address and ABI in TokenStaking.json");
                    return [3 /*break*/, 13];
                case 10: return [4 /*yield*/, deployments.deploy("TokenStaking", {
                        from: deployer,
                        args: tokenStakingConstructorArgs,
                        log: true,
                    })];
                case 11:
                    TokenStaking = _a.sent();
                    tokenStakingAddress = TokenStaking.address;
                    return [4 /*yield*/, execute("TokenStaking", { from: deployer }, "initialize")];
                case 12:
                    _a.sent();
                    log("Initialized TokenStaking.");
                    _a.label = 13;
                case 13:
                    if (!hre.network.tags.tenderly) return [3 /*break*/, 15];
                    return [4 /*yield*/, hre.tenderly.verify({
                            name: "TokenStaking",
                            address: tokenStakingAddress,
                        })];
                case 14:
                    _a.sent();
                    _a.label = 15;
                case 15: return [2 /*return*/];
            }
        });
    });
};
exports.default = func;
func.tags = ["TokenStaking"];
func.dependencies = [
    "T",
    "KeepTokenStaking",
    "NuCypherStakingEscrow",
    "VendingMachineKeep",
    "VendingMachineNuCypher",
    "KeepStake",
    "MintT",
];
